File: Sorting/merge.py
import sys

n = int(sys.stdin.readline())
arr = list(map(int, sys.stdin.readline().split()))

def merge(left, mid, right):
    i = left
    j = mid+1
    k = left
    saved = [0 * i for i in range(n)]

    while i <= mid and j <= right:
        # Search two arr and save small one to sorted array(saved)
        if arr[i] <= arr[j]:
            saved[k] = arr[i]
            k += 1
            i += 1
        else:
            saved[k] = arr[j]
            k += 1
            j += 1
    
    # Save remand one to sorted array(saved)
    if i > mid:
        for l in range(j, right+1):
            saved[k] = arr[l]
            k += 1
            l += 1
    else:
        for l in range(i, mid+1):
            saved[k] = arr[l]
            k += 1
            l += 1
    
    # recall sorted saved array to main array
    for l in range(left, right+1):
        arr[l] = saved[l]

def divide(left, right):
    if left < right:
        mid = (left+right)//2
        divide(left, mid)
        divide(mid+1, right)
        merge(left, mid, right)

divide(0, n-1)

print(arr)

File: Sorting/java/src/Merge.java

public class Merge {
    int[] arr;

    public Merge (int[] arr){
        this.arr = arr;
    }

    public int[] sort(){
        divide(0, arr.length-1);

        return arr;
    }

    private void divide(int left, int right) {
        if (left < right) {
            int mid = (left + right) / 2;
            divide(left, mid);
            divide(mid+1, right);
            merge(left, mid, right);
        }
    }

    private void merge(int left, int mid, int right) {
        int[] temporal = new int[right+1];
        int lpoint = left, rpoint = mid+1;
        int i = left;

        while (lpoint <= mid && rpoint <= right) {
            temporal[i++] = arr[lpoint] <= arr[rpoint] ? arr[lpoint++] : arr[rpoint++];
        }

        while (lpoint <= mid){
            temporal[i++] = arr[lpoint++];
        }

        while (rpoint <= right) {
            temporal[i++] = arr[rpoint++];
        }

        for (int l = left; l <= right; l++) {
            arr[l] = temporal[l];
        }
    }
}
